pkg/controller/daemon_controller.go

```diff
@@ -0,0 +1,432 @@
+				return dm.kubeClient.DaemonControllers(api.NamespaceAll).List(labels.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.DaemonControllers(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.DaemonController{},
+		FullDaemonControllerResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(old interface{}) {
+				glog.Infoln("Adding dc")
+				dm.enqueueController(old)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				glog.Infoln("Updating dc")
+				dm.enqueueController(cur)
+			},
+			DeleteFunc: func(old interface{}) {
+				glog.Infoln("Deleting dc")
+				dm.enqueueController(old)
+			},
+		},
+	)
+	// Watch for creation/deletion of pods. The reason we watch is that we don't want a daemon controller to create/delete
+	// more pods until all the effects (expectations) of a daemon controller's create/delete have been observed.
+	dm.podStore.Store, dm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		FullDaemonPodResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addPod,
+			UpdateFunc: dm.updatePod, // TODO: add update function, and call it
+			DeleteFunc: dm.deletePod,
+		},
+	)
+	// Watch for new nodes or updates to nodes - daemons are launched on new nodes, and possibly when labels on nodes change,
+	dm.nodeStore.Store, dm.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Nodes().Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Node{},
+		FullNodeResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addNode,
+			UpdateFunc: dm.updateNode,
+			DeleteFunc: func(node interface{}) {},
+		},
+	)
+	dm.syncHandler = dm.syncDaemonController
+	return dm
+}
+
+func (dm *DaemonManager) Run(workers int, stopCh <-chan struct{}) {
+	go dm.dcController.Run(stopCh)
+	go dm.podController.Run(stopCh)
+	go dm.nodeController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dm.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down Daemon Controller Manager")
+	dm.queue.ShutDown()
+}
+
+func (dm *DaemonManager) worker() {
+	for {
+		func() {
+			key, quit := dm.queue.Get()
+			if quit {
+				return
+			}
+			defer dm.queue.Done(key)
+			err := dm.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing daemon controller: %v", err)
```

Comment:
Please print the key in this error message so that log readers will have some idea of what dc is having issues.